### PR TITLE
Welcome page builds on npm compile and is loadable through the AWS Context Menu

### PR DESCRIPTION
**Pull request by bryceitoc9**
## Description
Created a welcome page that can be loaded through the AWS Explorer context menu
* Loads content from `extension-readme.md` in a webview
* HTML built through `npm run createWelcomePage` (accessible through `npm run compile`)
* When the extension is loaded, image paths need to be formatted like the following:
  * old: ``
    * We actually create the file `welcomePage.html` file with images loaded as `` to remove ambiguity when replacing the text to match below.
  * new: ``

## Motivation and Context
We need a welcome experience for VS Code. This is a start.

## Related Issue(s)
https://github.com/aws/aws-toolkit-vscode/issues/610

## Testing
Tests pass and you can manually open the page.

## Screenshots (if appropriate)
welcome-menu
welcome-page

## Checklist

- [x] I have read the **README** document
- [x] I have read the **CONTRIBUTING** document
- [x] My code follows the code style of this project
- [x] I have added tests to cover my changes
- [x] All new and existing tests passed
- [ ] A short description of the change has been added to the **CHANGELOG**
  - Will add a changelog entry upon completing https://github.com/aws/aws-toolkit-vscode/issues/611

## License

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: build-scripts/createWelcomePage.js [hunk 1]
```diff
@@ -0,0 +1,27 @@
+'use strict';
```
Comment (mpiroc) [inline build-scripts/createWelcomePage.js]: This (and other build scripts, in a separate change) needs a copyright header.

We should also probably convert our build-scripts to typescript, but that's out of scope for this change.

File: package.json [hunk 2]
```diff
@@ -305,9 +315,10 @@
         "vscode:prepublish": "npm run recompile",
         "bundleDeps": "node ./build-scripts/bundleDeps.js",
         "clean": "node ./build-scripts/clean.js out",
+        "createWelcomePage": "node ./build-scripts/createWelcomePage.js",
         "reset": "node ./build-scripts/clean.js out node_modules && npm install",
         "copyNonCodeFiles": "node ./build-scripts/copyNonCodeFiles.js",
-        "compile": "tsc -p ./ && npm run lint && npm run bundleDeps && npm run copyNonCodeFiles",
+        "compile": "tsc -p ./ && npm run lint && npm run bundleDeps && npm run copyNonCodeFiles && npm run createWelcomePage",
```
Comment (mpiroc) [inline package.json]: Can you add a script (something like `prepare` or `prebuild`) that runs `bundleDeps`, `copyNonCodeFiles` and `createWelcomePage`, then run the new script from both `compile` and `watch`? Otherwise, running `npm i` and `npm watch` directly after cloning will result in a broken build.

File: src/shared/extensionUtilities.ts [hunk 3]
```diff
@@ -69,3 +69,23 @@ export function safeGet<O, T>(obj: O | undefined, getFn: (x: O) => T): T | undef
...
+ * Useful for converting HTML images to webview-usable images
+ *
+ * @param extPath Extension path (from extension context)
+ * @param fileText Text of the file to scan
+ * @param token Token (string or regex) to search for (default: `/!!EXTENSIONROOT!!/g`)
+ */
+export function convertPathTokensToPath(
+    extPath: string,
+    fileText: string,
+    token: RegExp | string = /!!EXTENSIONROOT!!/g
```
Comment (mpiroc) [inline src/shared/extensionUtilities.ts]: Nit: If strings will just be converted (unescaped) to regexes anyway, suggest to remove `string` from the type union.

Also, minor naming suggestions:

```typescript
export function resolveResourcePaths(
    basePath: string,
    markdown: string,
    pattern: Regex = /!!EXTENSIONROOT!!/g
): string { ... }
```

Comment (bryceitoc9) [reply src/shared/extensionUtilities.ts]: `markdown` is inaccurate because it's already been converted to HTML at this point. Will switch to `text` since this can be applied to other things as well.

File: build-scripts/createWelcomePage.js (hunk 1 above)
Comment (bryceitoc9) [reply build-scripts/createWelcomePage.js]: Let's just add a separate fast-follow commit with the headers. That way, we don't have to worry about this potentially becoming inconsistent.

File: package.json (hunk 2 above)
Comment (bryceitoc9) [reply package.json]: ✔️

File: .gitignore [hunk 4]
```diff
@@ -6,3 +6,4 @@ node_modules
 **/.DS_Store
 .idea
 telemetryCache
+welcomePage.html
```
Comment (zhangzhx) [inline .gitignore]: Curiously, why don't you commit this file?

Comment (bryceitoc9) [reply .gitignore]: At this point, this file shouldn't be hand-edited. It should only be generated from the `extension-readme.md` file.

Thinking that we'll want to keep this in the future even if we move away from `extension-readme.md` as the base since Markdown is easier to modify...the only reason why we're transpiling to HTML is because VS Code can't render Markdown without manually opening a preview window.

Comment (zhangzhx) [reply .gitignore]: Gotcha, thanks for the explain.

File: package.json [hunk 5]
```diff
@@ -361,6 +373,7 @@
         "js-yaml": "^3.13.1",
         "jsonc-parser": "^2.0.2",
         "lodash": "^4.17.11",
+        "marked": "^0.6.2",
```
Comment (awschristou) [inline package.json]: Does `marked` belong in dependencies or devDependencies?

File: package.json [hunk 6]
```diff
@@ -305,11 +315,13 @@
...
         "clean": "node ./build-scripts/clean.js out",
+        "createWelcomePage": "node ./build-scripts/createWelcomePage.js",
         "reset": "node ./build-scripts/clean.js out node_modules && npm install",
         "copyNonCodeFiles": "node ./build-scripts/copyNonCodeFiles.js",
-        "compile": "tsc -p ./ && npm run lint && npm run bundleDeps && npm run copyNonCodeFiles",
+        "buildScripts": "npm run bundleDeps && npm run copyNonCodeFiles && npm run createWelcomePage",
+        "compile": "tsc -p ./ && npm run lint && npm run buildScripts",
         "recompile": "npm run clean && npm run compile",
-        "watch": "npm run copyNonCodeFiles && tsc -watch -p ./",
+        "watch": "npm run buildScripts && tsc -watch -p ./",
```
Comment (awschristou) [inline package.json]: This now runs `npm run bundleDeps`. Was this change intentional, and is there a negative impact in watch iterations?

File: src/extension.ts [hunk 7]
```diff
@@ -148,6 +150,23 @@ export async function activate(context: vscode.ExtensionContext) {
...
+        vscode.commands.registerCommand(
+            'aws.welcome',
+            async () => {
+                const view = vscode.window.createWebviewPanel(
+                    'html',
+                    localize('AWS.command.welcome.title', 'AWS Toolkit - Welcome'),
+                    vscode.ViewColumn.Active
+                )
+
+                const fileText = convertPathTokensToPath(
```
Comment (awschristou) [inline src/extension.ts]: Consider producing `fileText` prior to creating `view` in case an error occurs.

File: src/extension.ts [hunk 8]
```diff
@@ -148,6 +150,23 @@ export async function activate(context: vscode.ExtensionContext) {
             () => { vscode.env.openExternal(vscode.Uri.parse(githubUrl)) }
         )
 
+        vscode.commands.registerCommand(
+            'aws.welcome',
+            async () => {
+                const view = vscode.window.createWebviewPanel(
```
Comment (awschristou) [inline src/extension.ts]: Prefer to make this a standalone function, then provide the function to the registerCommand call.
The `activate` method needs cleanup, but we can manage the growth in line count.

File: src/shared/extensionUtilities.ts [hunk 9]
```diff
@@ -69,3 +69,19 @@ export function safeGet<O, T>(obj: O | undefined, getFn: (x: O) => T): T | undef
...
+
+/**
+ * Utility function to search for tokens in a string and convert them to relative paths parseable by VS Code
+ * Useful for converting HTML images to webview-usable images
+ *
+ * @param basePath Extension path (from extension context)
+ * @param text Text to scan
+ * @param pattern Regex to search for (default: `/!!EXTENSIONROOT!!/g`)
+ */
+export function convertPathTokensToPath(
```
Comment (awschristou) [inline src/shared/extensionUtilities.ts]: The parameter ordering is not intuitive.
Follow a more conventional "replace" signature: convertXyz(inputText, pattern, replaceWith).

Examples: [JS](https://www.w3schools.com/jsref/jsref_replace.asp), [C#](https://docs.microsoft.com/en-us/dotnet/api/system.string.replace?view=netframework-4.8#System_String_Replace_System_String_System_String_)

File: src/test/shared/extensionUtilities.test.ts [hunk 10]
```diff
@@ -23,4 +23,25 @@ describe('extensionUtilities', () => {
             assert.strictEqual(safeGet(undefined as Blah | undefined, x => x.someProp), undefined)
         })
     })
+
+    describe('convertPathTokensToPath', () => {
```
Comment (awschristou) [inline src/test/shared/extensionUtilities.test.ts]: Please add a test where the input text does not contain the token of interest

File: package.json (hunk 6 above)
Comment (bryceitoc9) [reply package.json]: See: https://github.com/aws/aws-toolkit-vscode/pull/619#discussion_r294525947

File: package.json (hunk 5 above)
Comment (bryceitoc9) [reply package.json]: Haven't tried devDependencies, I'll see if that works.

Comment (bryceitoc9) [reply package.json]: Looks like moving + `npm clean` + `npm install` + `npm run recompile` works. Leaving it in `devDependencies`.

File: src/shared/extensionUtilities.ts (hunk 9 above)
Comment (bryceitoc9) [reply src/shared/extensionUtilities.ts]: So the tricky part here is the default regex as a parameter...that only really works if that's in the final position.

Should I require that the caller provides this? At the very least, I can swap to `(inputText, replaceWith, pattern)`

File: src/shared/extensionUtilities.ts [hunk 11]
```diff
@@ -69,3 +73,53 @@ export function safeGet<O, T>(obj: O | undefined, getFn: (x: O) => T): T | undef
...
+    page: string = 'welcomePage.html'
+): Promise<vscode.WebviewPanel | void> {
+    let html: string | undefined
+    try {
+        html = convertExtensionRootTokensToPath(
+            await readFileAsString(path.join(context.extensionPath, page)),
+            context.extensionPath
+        )
+        if (!html) {
+            throw new Error()
```
Comment (awschristou) [inline src/shared/extensionUtilities.ts]: 1. Is anything logged indicating that the path could not be found (or any other error that may occur)?

2. A command will trigger this, usually through explicit request from the user. We should put a message in the output channel so they know nothing is going to happen. Outputting a message to the user is not the responsibility of this method, perhaps the method should not eat the error, and the command code that calls this method can catch+message

File: src/test/shared/extensionUtilities.test.ts (hunk 10 above)
Comment (bryceitoc9) [reply src/test/shared/extensionUtilities.test.ts]: ✔️ Added to the `createWelcomeWebview` tests

File: src/shared/extensionUtilities.ts (hunk 9 above)
Comment (bryceitoc9) [reply src/shared/extensionUtilities.ts]: ✔️ Function is now private and has the token hardcoded. Now fits the pattern of `inputText, replaceWith`

File: src/extension.ts (hunk 8 above)
Comment (bryceitoc9) [reply src/extension.ts]: ✔️

File: src/extension.ts (hunk 7 above)
Comment (bryceitoc9) [reply src/extension.ts]: ✔️

File: package.json (hunk 5 above)
Comment (bryceitoc9) [reply package.json]: ✔️

File: src/test/shared/extensionUtilities.test.ts [hunk 12]
```diff
@@ -23,4 +29,54 @@ describe('extensionUtilities', () => {
...
+        const context = new FakeExtensionContext()
+        let filepath: string | undefined
+
+        before(async () => {
+            const tempDir = await filesystemUtilities.makeTemporaryToolkitFolder()
+            context.extensionPath = tempDir
+        })
+
+        afterEach(async () => {
+            if (filepath && await filesystemUtilities.fileExists(filepath)) {
```
Comment (awschristou) [inline src/test/shared/extensionUtilities.test.ts]: Why not create `tempDir` and set extensionPath before _each_ test, then unconditionally delete tempDir after each test? Many tests do this on a per test basis. This will simplify the setup/teardown, and eliminate the risk of side effects. 

If you're concerned about extensionPath not referencing a valid location, it can be set to undefined after each test as well.

File: src/shared/extensionUtilities.ts (hunk 11 above)
Comment (bryceitoc9) [reply src/shared/extensionUtilities.ts]: No, what would be the most ideal place to surface this? A toast? Right now, the welcome page will only show on a user directly invoking this command, so immediate feedback might be best.

Comment (bryceitoc9) [reply src/shared/extensionUtilities.ts]: Also, this was largely done to mitigate https://github.com/aws/aws-toolkit-vscode/pull/619#discussion_r294975639

Comment (awschristou) [reply src/shared/extensionUtilities.ts]: I think the Output Channel would work. The reason we wouldn't place the notification directly in this method is that this method could be called from elsewhere, and it may not make sense in those contexts to surface a similar message. 

Re: The comment you're trying to mitigate, there appears to still be logic in the async handler. Could all of that be placed in its own method?

Comment (bryceitoc9) [reply src/shared/extensionUtilities.ts]: re: second point, yes, I can add that logic to `createWelcomeWebview`. In that case, I'll swap the name to `showWelcomeWebview` since that's what it'll be doing (or having a `showWelcomeWebview` that wraps `create` and does the explicit focus).

re: the first point, this is an error state that a user should never see, and if the user does see it, it would be in direct reaction to a user action (in its current state). I'm not sure if the output channel is enough feedback in a case like this...as a user, if I click something, I expect some sort of immediate, obvious visual feedback. I don't feel that an output channel that may already contain information is the best way to surface this type of failure. I can't think of any scenario that we'd call for the welcome page with absolutely no user input.

That being said, we should send any failures with this as a telemetry event. I'll look into that.

File: src/extension.ts [hunk 13]
```diff
@@ -148,6 +148,16 @@ export async function activate(context: vscode.ExtensionContext) {
             () => { vscode.env.openExternal(vscode.Uri.parse(githubUrl)) }
         )
 
+        vscode.commands.registerCommand(
```
[the remark this answers is not in the document]
Comment (bryceitoc9) [reply src/extension.ts]: I'm thinking we should hold off on #2 until we have a metric decided on in https://github.com/aws/aws-toolkit-vscode/issues/625 . Thoughts?

As of now, this matches the implementation for `View Documentation` and `View Source`.

[the remark this answers is not in the document]
Comment (awschristou) [reply src/extension.ts]: Sounds good, preferably 625 is a fast follow
